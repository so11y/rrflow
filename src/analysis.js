import destructuring from "@babel/plugin-transform-destructuring";
import { transform, packages } from "@babel/standalone";

function getRefNodeName(path) {
  const { types } = packages;
  switch (path.type) {
    case "Identifier":
      return path.node.name;
    case "MemberExpression": {
      let node = path.node;
      while (!types.isIdentifier(node)) {
        node = node.object;
      }
      return node.name;
    }
  }
  throw `不支持`;
}

function getExecutePathName(path) {
  const { generator, types } = packages;
  switch (path.type) {
    case "Identifier":
      return types.StringLiteral(path.node.name);
    case "MemberExpression": {
      return types.StringLiteral(generator.default(path.node).code);
    }
  }
  throw `不支持`;
}

function isLoopStatement(path) {
  const { types } = packages;
  const validates = [types.isForStatement, types.isWhileStatement];
  return validates.some((validate) => validate(path));
}

export function builder(code) {
  const { types, template, generator } = packages;
  const t = types;
  return transform(code, {
    plugins: [
      destructuring,
      {
        visitor: {
          Program: {
            enter(path, state) {
              state.scopeTrack = path.scope.generateUidIdentifier("scopeTrack");
              state.scopeHelper =
                path.scope.generateUidIdentifier("scopeHelper");
            },
            exit(path, state) {
              const scopeHelper = template.default(`
                const SCOPE_HELPER = __codeRecodeScope__;
              `)({
                SCOPE_HELPER: state.scopeHelper,
              });
              path
                .unshiftContainer("body", scopeHelper)
                .forEach((path) => path.skip());

              const exit = template.default(`SCOPE_HELPER.exit()`)({
                SCOPE_HELPER: state.scopeHelper,
              });

              path.pushContainer("body", exit).forEach((path) => path.skip());
            },
          },
          VariableDeclaration: {
            exit(path, state) {
              path.get("declarations").forEach((decPath) => {
                const tempCode = template.default`SCOPE_HELPER.variable(NAME,CODE)`;
                const newCode = tempCode({
                  NAME: types.StringLiteral(decPath.get("id.name").node),
                  CODE: decPath.get("init").node,
                  SCOPE_HELPER: state.scopeHelper,
                }).expression;
                decPath.set("init", newCode);
              });
              path.skip();
            },
          },
          CallExpression: {
            exit(path, state) {
              const tempCode = template.default`SCOPE_HELPER.execute(EXPRESSION,()=> FN)`;
              path.replaceWith(
                tempCode({
                  FN: path.node,
                  EXPRESSION: getExecutePathName(path.get("callee")),
                  SCOPE_HELPER: state.scopeHelper,
                }).expression
              );
              path.skip();
            },
          },
          ReturnStatement: {
            exit(path, state) {
              const tempCode = template.default`
              return SCOPE_TRACK.drop(NODE,true)
              `;
              path.replaceWith(
                tempCode({
                  SCOPE_TRACK: state.scopeTrack,
                  NODE: path.get("argument").node,
                })
              );
              path.skip();
            },
          },
          BreakStatement: {
            exit(path, state) {
              const nodes = [
                template.default.ast(`${state.scopeTrack.name}.drop()`),
                path.node,
              ];
              const newPath = path.replaceWithMultiple(nodes);
              newPath.forEach((path) => path.skip());
            },
          },
          "ForStatement|WhileStatement|IfStatement": {
            enter(_, state) {
              if (state.forIndex === undefined) {
                state.forIndex = 0;
              } else {
                state.forIndex++;
              }
            },
            exit(path, state) {
              const scopeName = isLoopStatement(path) ? "for" : "if";
              const tempCode = template.default(
                `{
                /****** ${state.forIndex} ******/
                const SCOPE_TRACK = SCOPE_HELPER.createScope("${scopeName}");
                CODE
                SCOPE_TRACK.drop();
                /****** ${state.forIndex} ******/
               }`,
                {
                  preserveComments: true,
                }
              );

              const BlockStatement = tempCode({
                SCOPE_TRACK: state.scopeTrack,
                SCOPE_HELPER: state.scopeHelper,
                CODE: path.node,
              });
              path.replaceWith(BlockStatement);
              path.skip();
              state.forIndex--;
            },
          },
          "BinaryExpression|LogicalExpression": {
            exit(path, state) {
              if (path.key === "test") {
                const code = generator.default(path.node).code;
                const testAst = template.default(
                  `SCOPE_TRACK.test(EXPRESSION,CODE)`
                )({
                  CODE: path.node,
                  EXPRESSION: types.StringLiteral(code),
                  SCOPE_TRACK: state.scopeTrack,
                });
                path.replaceWith(testAst);
                path.skip();
              }
            },
          },
          "AssignmentExpression|UpdateExpression": {
            exit(path, state) {
              const leftPath = path.isUpdateExpression()
                ? path.get("argument")
                : path.get("left");
              let refName = getRefNodeName(leftPath);
              const tempCode = template.default(
                `SCOPE_HELPER.setVariable("REF_NAME",(NODE,REF_NAME))`
              );
              path.replaceWith(
                tempCode({
                  REF_NAME: refName,
                  NODE: path.node,
                  SCOPE_HELPER: state.scopeHelper,
                })
              );
              path.skip();
            },
          },
          "FunctionDeclaration|ArrowFunctionExpression|ObjectMethod": {
            exit(path, state) {
              const trackHelper = template.default(`
                  const SCOPE_TRACK = SCOPE_HELPER.getCurrentScopeHelper()
                  `)({
                SCOPE_TRACK: state.scopeTrack,
                SCOPE_HELPER: state.scopeHelper,
              });
              path
                .get("body")
                .unshiftContainer("body", trackHelper)
                .forEach((path) => path.skip());
            },
          },
        },
      },
    ],
  }).code;
}
