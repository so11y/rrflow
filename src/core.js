import { cloneDeep, isFunction } from "lodash-es";
export function __codeRecodeScope__() {
  const scopeHelperCache = new WeakMap();
  const scopeHelper = {
    _scope: [],
    _isExit: false,
    currentScope: null,
    createScope(name) {
      const newScope = {
        name,
        parent: scopeHelper.currentScope,
        /** dyn keys */
        /**
         *   returned: undefined,
         *   var:{}
         *   function:{},
         *   children:[]
         *   dirtyChange:{} 修改了哪一些值
         */
      };
      if (scopeHelper.currentScope) {
        if (!scopeHelper.currentScope.children) {
          scopeHelper.currentScope.children = [];
        }
        scopeHelper.currentScope.children.push(newScope);
      }
      scopeHelper._scope.push(newScope);
      scopeHelper.currentScope = newScope;
      scopeHelperCache.set(newScope, {
        test(expression, value) {
          const scope = newScope;
          if (!scope.test) {
            scope.test = [];
          }
          scope.test.push({
            expression,
            value,
          });
          return value;
        },
        drop: (value, exit) => {
          const isDrop = scopeHelper._scope.findIndex(
            (scope) => scope === newScope
          );
          if (isDrop === scopeHelper._scope.length - 1) {
            const dropScope = scopeHelper._scope.pop();
            scopeHelper.currentScope = dropScope.parent;
            if (exit) {
              dropScope.returned = cloneDeep(value);
            }
          }
          if (exit && scopeHelper.currentScope.name === "global") {
            scopeHelper.exit();
          }
        },
      });
      return scopeHelperCache.get(newScope);
    },
    getCurrentScopeHelper() {
      return scopeHelperCache.get(scopeHelper.currentScope);
    },
    execute(name, fn) {
      const scope = scopeHelper.createScope("fn");
      let value = fn();
      scope.name = name;
      scope.returned = cloneDeep(value);
      scope.drop(value);
      return value;
    },
    variable(key, value) {
      const isValueFunction = isFunction(value);
      const visitorKey = isValueFunction ? "function" : "var";
      if (!scopeHelper.currentScope[visitorKey]) {
        scopeHelper.currentScope[visitorKey] = {};
      }
      scopeHelper.currentScope[visitorKey][key] = {
        value,
        recode: [],
      };
      return value;
    },
    setVariable(key, value) {
      let scope = scopeHelper.currentScope;
      while (!scope.var || !Reflect.has(scope.var, key)) {
        scope = scope.parent;
      }
      scope.var[key].recode.push({
        value,
        ordValue: cloneDeep(scope.var[key].value),
        changeScope: scopeHelper.currentScope,
      });
      scope.var[key].value = value;
    },
    exit() {
      let globalScope = scopeHelper.currentScope;
      if (!scopeHelper._isExit) {
        while (globalScope.name !== "global") {
          scopeHelperCache.get(globalScope).drop();
          globalScope = globalScope.parent;
        }
      }
      return globalScope;
    },
  };
  scopeHelper.createScope("global");
  return scopeHelper;
}

const runInSandBox = (sandbox) => {
  const context = new Proxy(sandbox, {
    has: () => true,
    get(target, key) {
      if (key === Symbol.unscopables) return undefined;
      return target[key];
    },
  });
  return (rawCode) => {
    const code = new Function(
      "sandbox",
      `with (sandbox) {
      ${rawCode}
     }`
    );
    code(context);
    return context;
  };
};

export function sandBox(code) {
  return runInSandBox({
    Math,
    console,
    __codeRecodeScope__: __codeRecodeScope__(),
  })(code);
}
