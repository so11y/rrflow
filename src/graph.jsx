import { Graph } from "@antv/x6";
import { DagreLayout } from "@antv/layout";
import { onMount, createEffect } from "solid-js";

function createRenderNode(recode) {
  const nodes = [];
  const stack = [recode];
  const edges = [];
  while (stack.length) {
    const current = stack.pop();
    current.shape = "rect";
    current.id = current.id.toString();
    current.size = {
      width: 100,
      height: 40,
    };
    current.label = current.executeName ?? current.name;
    if (current.parent) {
      const where = current.where;
      edges.push({
        source: current.parent.id,
        target: current.id,
        labels: [
          {
            attrs: {
              label: {
                text: where
                  ? `条件:${where.expression},结果:${where.value}`
                  : "",
              },
            },
          },
        ],
      });
    }
    nodes.push(current);
    if (current.children) {
      stack.push(...current.children);
    }
  }
  return {
    nodes,
    edges,
  };
}

export default function graph(props) {
  let container;
  let graphImpl;
  onMount(() => {
    graphImpl = new Graph({
      container,
      grid: true,
      background: {
        color: "#F2F7FA",
      },
    });
  });

  createEffect(() => {
    console.log("recode =", props.recode);
    if (props.recode) {
      const layout = new DagreLayout({
        type: "dagre", // 布局类型
        rankdir: "TB", // 布局的方向
        nodesep: "30", // 节点间距
        ranksep: "0", // 层间距
      });
      console.log(
        createRenderNode(props.recode),'-'
      );
      const model = layout.layout(createRenderNode(props.recode));
      graphImpl.fromJSON(model);
      graphImpl.centerContent();
      graphImpl.zoomToFit({
        padding: {
          left: 10,
          right: 10,
        },
      });
    }
  });

  return <div className="flex-auto" ref={container} id="container"></div>;
}
