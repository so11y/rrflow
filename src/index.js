function builder(code) {
  const { types, template, traverse, generator } = Babel.packages;
  const transformed = Babel.transform(code, {
    plugins: [
      {
        visitor: {
          VariableDeclaration: {
            exit(path) {
              path.get("declarations").map((decPath) => {
                const temp = `${path.get('kind').node} xx  = variable(xx,xx)`;
                const tempCode = template.default`${temp}`;
                const newCode = tempCode({
                  // NAME: decPath.get("id").node,
                  // CODE: decPath.get("init").node,
                });

                console.log(newCode.expression);
                // decPath.set("init", newCode);
                return newCode.expression
              });
              path.skip();
            },
          },
          CallExpression: {
            exit(path) {
              const tempCode = template.default`callFn(()=> FN)`;
              path.replaceWith(
                tempCode({
                  FN: path.node,
                }).expression
              );
              console.log(456);
              path.skip();
            },
          },
          ForStatement: {
            enter(_, state) {
              if (state.forIndex === undefined) {
                state.forIndex = 0;
              } else {
                state.forIndex++;
              }
            },
            exit(path, state) {
              const tempCode = template.default`
            {
              const {drop}= createScope("for");
              CODE
              drop();
            }
              `;
              path.replaceWith(
                tempCode({
                  CODE: path.node,
                })
              );
              path.skip();
              state.forIndex--;
            },
          },
        },
      },
    ],
  }).code;

  console.log(transformed, "-");
}

builder(`

const {a} = aa;
`);

// let a1 = function(){
//   let index1 = 0;
//   let index2 = 0;
// }
//   a1();

//     for (
//     let index = 0;
//     index < 10;
// index++
//    ) {

//      let index1 = 0;
//    }

// var xx = {a:1};
// let a = vaa("a",xx,1)
// let b = vaa("b",xx,2)

// path.get("declarations").forEach((decPath) => {
//   const tempCode = template.default`variable(NAME,CODE)`;
//   console.log(decPath.get("id").node, "--");
//   const newCode = tempCode({
//     NAME: decPath.get("id").node,
//     CODE: decPath.get("init").node,
//   }).expression;
//   console.log(generator.default((decPath.get("id").node)));
//   // decPath.set("init", newCode);
// });
// path.skip();
