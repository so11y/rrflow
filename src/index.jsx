import { render } from 'solid-js/web'
import App from './app.jsx'
import 'virtual:uno.css'
import "./index.scss"

const root = document.getElementById('root')

render(() => <App />, root)
