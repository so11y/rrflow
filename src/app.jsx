import { createSignal, onMount } from 'solid-js'
import { builder } from "../flow/analysis";
import { sandBox, } from "../flow/core";
import Graph from "./graph"


function App() {
  const [code, setCode] = createSignal(`function sumOfOdds(n) {
  let sum = 0;
  for (let i = 1; i <= n; i++) {
    if (i === 2) {
      console.log(5);
    } else {
      console.log(6);
    }
  }
  return sum;
}
sumOfOdds(10);`)
  const [translateCode, setTranslateCode] = createSignal("");
  const [recode, setRecode] = createSignal(null)
  function translate() {
    setTranslateCode(builder(code()))
    try {
      setRecode(sandBox(translateCode()).__codeRecodeScope__
        .currentScope)
    } catch (error) {
      console.error(error)
    }

  }
  onMount(translate)

  return (
    <div className='min-w-screen min-h-screen  flex p-8px  box-border'>
      <div className='w-40% flex flex-col'>
        <button onClick={translate}>编译</button>
        <textarea className='w-full flex-auto resize-none' value={code()} onInput={(e) => setCode(e.currentTarget.value)}></textarea>
        <textarea readOnly className='flex-auto w-full' value={translateCode()} />
      </div>
      <Graph recode={recode()} />
    </div>
  )
}

export default App
