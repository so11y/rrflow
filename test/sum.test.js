import { expect, test } from "vitest";
import { builder } from "../src/analysis";

test("Sum of Odds", () => {
  const code = builder(`
    function sumOfOdds(n) {
      let sum = 0;
      for (let i = 1; i <= n; i++) {
        if (i % 2 !== 0) {
          sum += i;
        }
      }
      return sum;
    }
    console.log(sumOfOdds(10)); // 输出 25
    `);

  expect(code).toMatchSnapshot();
});
