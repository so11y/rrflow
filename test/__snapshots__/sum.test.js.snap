// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sum of Odds 1`] = `
"function sumOfOdds(n) {
  let sum = _scopeHelper.variable("sum", 0);
  {
    /****** 0 ******/
    const _scopeTrack = _scopeHelper.createScope("for");
    for (let i = _scopeHelper.variable("i", 1); i <= n && _scopeTrack.track("i <= n"); _scopeHelper.setVariable("i", (i++, i))) {
      {
        /****** 1 ******/
        const _scopeTrack = _scopeHelper.createScope("if");
        if (i % 2 !== 0 && _scopeTrack.track("i % 2 !== 0")) {
          _scopeHelper.setVariable("sum", (sum += i, sum));
        }
        _scopeTrack.drop();
        /****** 1 ******/
      }
    }
    _scopeTrack.drop();
    /****** 0 ******/
  }
  return _scopeTrack.drop(sum);
}
_scopeHelper.execute(() => console.log(_scopeHelper.execute(() => sumOfOdds(10)))); // 输出 25"
`;
